### ISSUE TITLE
lanfactory drop torch

### ISSUE THREAD

**Issue by cpaniaguam** [lnccbrown/LANfactory#30]
There are a few modules where `torch` is used and at least one test relies on it.

@AlexanderFengler can you provide some guidelines?

<details>
    
     
    118 results - 7 files
    
    lanfactory/onnx/transform_onnx.py:
       4  
       5: import torch
       6: from lanfactory.trainers.torch_mlp import TorchMLP
       7  
       8: """This module contains the function to transform Torch/Jax models to ONNX format.
       9  Can be run as a script.
    
      19      """
      20:     Transforms a TorchMLP model to ONNX format.
      21  
    
      35  
      36:     mynet = TorchMLP(
      37          network_config=network_config_mlp,
    
      42      mynet.load_state_dict(
      43:         torch.load(state_dict_file, map_location=torch.device("cpu")),
      44      )
      45  
      46:     x = torch.randn(1, input_shape, requires_grad=True)
      47:     torch.onnx.export(mynet, x, output_onnx_file)
      48  
    
      50  if __name__ == "__main__":
      51:     parser = argparse.ArgumentParser(description="Convert a TorchMLP model to ONNX format.")
      52      parser.add_argument("network_config_file", help="Path to the network configuration file (pickle).")
    
    lanfactory/trainers/__init__.py:
       1: from .torch_mlp import DatasetTorch, TorchMLP
       2: from .torch_mlp import ModelTrainerTorchMLP, LoadTorchMLPInfer, LoadTorchMLP
       3  from .jax_mlp import MLPJaxFactory, MLPJax, ModelTrainerJaxMLP
    
       5  __all__ = [
       6:     "DatasetTorch",
       7:     "TorchMLP",
       8:     "ModelTrainerTorchMLP",
       9:     "LoadTorchMLPInfer",
      10:     "LoadTorchMLP",
      11      "MLPJaxFactory",
    
    lanfactory/trainers/jax_mlp.py:
      244                  The MLPJax model to be trained.
      245:             train_dl (torch.utils.data.DataLoader):
      246                  The training data loader.
      247:             valid_dl (torch.utils.data.DataLoader):
      248                  The validation data loader.
    
    lanfactory/trainers/torch_mlp.py:
        8  
        9: import torch
       10: import torch.nn as nn
       11: import torch.optim as optim
       12: import torch.nn.functional as F
       13: from torch.utils.data import DataLoader
       14  
    
       21  
       22: """This module contains the classes for training TorchMLP models."""
       23  
       24  
       25: class DatasetTorch(torch.utils.data.Dataset):
       26:     """Dataset class for TorchMLP training.
       27  
    
       53          label_key: str = "labels",
       54:         out_framework: str = "torch",
       55      ) -> None:
    
      150  
      151: class TorchMLP(nn.Module):
      152:     """TorchMLP class.
      153  
    
      173      ) -> None:
      174:         super(TorchMLP, self).__init__()
      175  
    
      193          self.activations: dict[str, nn.Module] = {
      194:             "relu": torch.nn.ReLU(),
      195:             "tanh": torch.nn.Tanh(),
      196:             "sigmoid": torch.nn.Sigmoid(),
      197          }
    
      231      # Define forward pass
      232:     def forward(self, x: torch.Tensor) -> torch.Tensor:
      233          """Forward pass through network.
    
      236          ---------
      237:             x (torch.Tensor):
      238                  Input tensor.
    
      241          -------
      242:             torch.Tensor:
      243                  Output tensor.
    
      249          elif self.train_output_type == "logits":
      250:             return -torch.log(
      251:                 (1 + torch.exp(-self.layers[-1](x)))
      252              )  # log ( 1 / (1 + exp(-x))), where x = log(p / (1 - p))
    
      256  
      257: class ModelTrainerTorchMLP:
      258      def __init__(
      259          self,
      260:         model: TorchMLP,
      261          train_config: dict | str,
    
      267      ) -> None:
      268:         """Class to train Torch Models.
      269          Arguments
    
      272                  Training configuration.
      273:             model (TorchMLP):
      274:                 TorchMLP model.
      275:             train_dl (DatasetTorch):
      276                  Training dataloader.
      277:             valid_dl (DatasetTorch):
      278                  Validation dataloader.
    
      285          """
      286:         torch.backends.cudnn.benchmark = True
      287:         self.dev: torch.device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
      288:         print("Torch Device: ", self.dev)
      289          if train_config is None:
    
      302  
      303:         self.model: TorchMLP = model.to(self.dev)
      304          self.allow_abs_path_folder_generation: bool = allow_abs_path_folder_generation
    
      506  
      507:                 pred: torch.Tensor = self.model(xb)
      508:                 loss: torch.Tensor = self.loss_fun(pred, yb)
      509  
    
      531              # self.model.eval()
      532:             with torch.no_grad():
      533:                 val_loss: torch.Tensor = (
      534                      sum(self.loss_fun(self.model(xb.to(self.dev)), yb.to(self.dev)) for xb, yb in self.valid_dl)
    
      561              print("Saving training history")
      562:             training_history_path: str = full_path + "_torch_training_history.csv"
      563              pd.DataFrame(training_history).to_csv(training_history_path)
    
      568              print("Saving model state dict")
      569:             train_state_path: str = full_path + "_train_state_dict_torch.pt"
      570:             torch.save(
      571                  self.model.state_dict(),
    
      599              print("Saving model to ONNX format")
      600:             onnx_path: str = full_path + "_torch_model.onnx"
      601              # Put model in eval mode
      602              self.model.eval()
      603:             torch.onnx.export(
      604                  self.model,
      605:                 torch.randn(1, self.model.input_shape).to(self.dev),
      606                  onnx_path,
    
      622  
      623: class LoadTorchMLPInfer:
      624:     """Class to load TorchMLP models for inference. (This
      625      was originally useful directly for application in the
    
      644      ) -> None:
      645:         torch.backends.cudnn.benchmark = True
      646:         self.dev = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
      647          self.model_file_path = model_file_path
    
      657  
      658:         self.net = TorchMLP(
      659              network_config=self.network_config,
    
      662          )
      663:         if not torch.cuda.is_available():
      664:             self.net.load_state_dict(torch.load(self.model_file_path, map_location=torch.device("cpu")))
      665          else:
      666:             self.net.load_state_dict(torch.load(self.model_file_path))
      667          self.net.to(self.dev)
    
      669  
      670:     @torch.no_grad()
      671:     def __call__(self, x: torch.Tensor) -> torch.Tensor:
      672          return self.net(x)
      673  
      674:     @torch.no_grad()
      675      def predict_on_batch(self, x: np.ndarray | None = None) -> np.ndarray:
    
      699          """
      700:         return self.net(torch.from_numpy(x).to(self.dev)).cpu().numpy()
      701  
      702  
      703: class LoadTorchMLP:
      704:     """Class to load TorchMLP models.
      705  
    
      720      ) -> None:
      721:         self.dev = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
      722          self.model_file_path = model_file_path
    
      732  
      733:         self.net = TorchMLP(
      734              network_config=self.network_config,
    
      737          )
      738:         if not torch.cuda.is_available():
      739:             self.net.load_state_dict(torch.load(self.model_file_path, map_location=torch.device("cpu")))
      740          else:
      741:             self.net.load_state_dict(torch.load(self.model_file_path))
      742          self.net.to(self.dev)
      743  
      744:     @torch.no_grad()
      745:     def __call__(self, x: torch.Tensor) -> torch.Tensor:
      746          return self.net(x)
      747  
      748:     @torch.no_grad()
      749      def predict_on_batch(self, x: np.ndarray | None = None) -> np.ndarray:
      750:         return self.net(torch.from_numpy(x).to(self.dev)).cpu().numpy()
    
    tests/conftest.py:
       3  import os
       4: import torch
       5  import jax
    
      54  @pytest.fixture
      55: def torch_device():
      56:     """Fixture to provide a PyTorch device (CPU/GPU)."""
      57:     return torch.device("cuda" if torch.cuda.is_available() else "cpu")
      58  
    
    tests/test_end_to_end.py:
        7  import jax.numpy as jnp
        8: import torch
        9  from .constants import (
    
      104      # Training dataset
      105:     jax_training_dataset = lanfactory.trainers.DatasetTorch(
      106          file_ids=file_list_,
    
      113  
      114:     jax_training_dataloader = torch.utils.data.DataLoader(
      115          jax_training_dataset,
    
      122      # Validation dataset
      123:     jax_validation_dataset = lanfactory.trainers.DatasetTorch(
      124          file_ids=file_list_,
    
      131  
      132:     jax_validation_dataloader = torch.utils.data.DataLoader(
      133          jax_validation_dataset,
    
    tests/test_fixtures.py:
       1  import numpy as np
       2: import torch
       3  
    
      13  
      14: def test_torch_device(torch_device):
      15:     """Test that the torch device fixture works."""
      16:     assert isinstance(torch_device, torch.device)
      17  
</details>

**Comment by AlexanderFengler** [comment]
@cpaniaguam closing this as result of recent conversations that made us realize we can't really drop torch very easily for now (if at all).
